pkl-config-java/pkl-config-java.gradle.kts

```diff
@@ -10,14 +10,18 @@ val pklCodegenJava: Configuration by configurations.creating
 val firstPartySourcesJars by configurations.existing
 
 val generateTestConfigClasses by tasks.registering(JavaExec::class) {
-  outputs.dir("build/testConfigClasses")
+  val outputDir = layout.buildDirectory.dir("testConfigClasses")
+  outputs.dir(outputDir)
   inputs.dir("src/test/resources/codegenPkl")
 
   classpath = pklCodegenJava
   mainClass.set("org.pkl.codegen.java.Main")
-  args("--output-dir", "build/testConfigClasses")
-  args("--generate-javadoc")
-  args(fileTree("src/test/resources/codegenPkl"))
+  argumentProviders.add(CommandLineArgumentProvider {
```

Comment:
Using an argument provider has no benefits here.
This should work fine:
```
args("--output-dir", outputDir.get().asFile)
args("--generate-javadoc")
args(fileTree("src/test/resources/codegenPkl"))
```

Reply:
Calling `outputDir.get()` is eager. So, this just lets us configure the task while being able to keep the laziness of the `buildDirectory` provider.

As far as I can tell, this is the only way to configure an exec task while keeping it lazy.

Reply:
And what's the benefit of keeping the laziness of the buildDirectory provider? Either my Gradle knowledge is outdated, or you have some misconceptions about this.

Reply:
According to their docs, the whole reason they switched from the eager `getBuildDir()` to this lazy API is to prevent ordering issues: https://docs.gradle.org/current/userguide/upgrading_version_8.html#deprecations_4

Seems like using this API eagerly kind of defeats the purpose; we'd still be vulnerable to ordering issues because we'd be configuring the task eagerly.

Note: I don't think we'd actually be vulnerable since we're not changing the build dir. But, in any case, it's good to try to be more idiomatic.

Reply:
Right, the `buildDir` ordering issue is irrelevant to Pkl and most other projects.

Gradle has been slowly migrating from `Foo` to `Property<Foo>`/`Provider<Foo>` for years.
For APIs that have been migrated, it makes sense to avoid `get()` at configuration time 
and use `property.set(provider)` instead of `property.set(foo)`.
This prevents ordering issues and enables task dependency tracking.
For the many APIs that haven't been migrated, using `get()` is normal and expected.

`JavaExec.args` hasn't been migrated to `ListProperty`, which would allow `args.add(outputDir)`.
Using an argument provider only to delay `outputDir.get()` isn't worth its weight.

If you'd like to dive deeper into this madness, I recommend reading through https://github.com/gradle/gradle/issues/4685.

Reply:
> For the many APIs that haven't been migrated, using `get()` is normal and expected.

I'd agree if there was no way to lazily configure the arguments. But since there is here, we'll just keep it this way. I don't think it's adding too much complexity to do it this way.

But yeah, it's pretty silly that they haven't provided a way to set `args` directly with lazy values yet.